### PR TITLE
Implementation of Process Controller

### PR DESCRIPTION
**Pull request by caleeli**
/api/1.0/processes?filter=ACTIVE&include=category,category.processes
```json
{
    "data": [{
            "type": null,
            "id": "f4e07138-b6cb-11e8-87a2-080027874bc2",
            "attributes": {
                "uuid": "f4e07138-b6cb-11e8-87a2-080027874bc2",
                "process_category_uuid": "f4d50b36-b6cb-11e8-a5ff-080027874bc2",
                "user_uuid": "f4d4ae02-b6cb-11e8-a4fd-080027874bc2",
                "description": "Tempora explicabo cupiditate dolorem nisi vel dolores. Dolorem voluptatem dolore quas laudantium esse quae quidem. Aut labore sed quas praesentium.",
                "name": "Aliquam distinctio quia.",
                "status": "INACTIVE",
                "created_at": "2018-09-12 20:39:33",
                "updated_at": "2018-09-12 20:39:33"
            },
            "relationships": {
                "category": {
                    "data": {
                        "type": null,
                        "id": "f4d50b36-b6cb-11e8-a5ff-080027874bc2"
                    }
                }
            }
        }, {
            "type": null,
            "id": "f4df5e24-b6cb-11e8-9c78-080027874bc2",
            "attributes": {
                "uuid": "f4df5e24-b6cb-11e8-9c78-080027874bc2",
                "process_category_uuid": "f4af3dd4-b6cb-11e8-96f4-080027874bc2",
                "user_uuid": "f4aedccc-b6cb-11e8-bff6-080027874bc2",
                "description": "Minus perspiciatis dolores adipisci occaecati alias labore ad reprehenderit. Voluptatum corrupti voluptatem doloribus placeat perferendis provident laudantium eos. In id vero ipsam voluptatem sed. A voluptatem sunt at dolor voluptatem voluptatibus quia sequi.",
                "name": "Autem omnis dolor culpa corporis.",
                "status": "INACTIVE",
                "created_at": "2018-09-12 20:39:33",
                "updated_at": "2018-09-12 20:39:33"
            },
            "relationships": {
                "category": {
                    "data": {
                        "type": null,
                        "id": "f4af3dd4-b6cb-11e8-96f4-080027874bc2"
                    }
                }
            }
        }, {
            "type": null,
            "id": "f4e0285e-b6cb-11e8-a376-080027874bc2",
            "attributes": {
                "uuid": "f4e0285e-b6cb-11e8-a376-080027874bc2",
                "process_category_uuid": "f4cb7710-b6cb-11e8-bded-080027874bc2",
                "user_uuid": "f4cb1824-b6cb-11e8-a414-080027874bc2",
                "description": "Voluptatum iusto eos officia distinctio. Laboriosam ut dolores ut expedita. Qui ullam impedit neque et aspernatur labore quia.",
                "name": "Earum itaque quidem.",
                "status": "INACTIVE",
                "created_at": "2018-09-12 20:39:33",
                "updated_at": "2018-09-12 20:39:33"
            },
            "relationships": {
                "category": {
                    "data": {
                        "type": null,
                        "id": "f4cb7710-b6cb-11e8-bded-080027874bc2"
                    }
                }
            }
        }, {
            "type": null,
            "id": "f4deca90-b6cb-11e8-b0c8-080027874bc2",
            "attributes": {
                "uuid": "f4deca90-b6cb-11e8-b0c8-080027874bc2",
                "process_category_uuid": "f49af20c-b6cb-11e8-9dc7-080027874bc2",
                "user_uuid": "f49a863c-b6cb-11e8-8e8e-080027874bc2",
                "description": "Sed eos porro et nemo harum. Aut blanditiis corporis non in rerum iusto. Quis omnis qui officiis distinctio.",
                "name": "Facere nobis et.",
                "status": "INACTIVE",
                "created_at": "2018-09-12 20:39:33",
                "updated_at": "2018-09-12 20:39:33"
            },
            "relationships": {
                "category": {
                    "data": {
                        "type": null,
                        "id": "f49af20c-b6cb-11e8-9dc7-080027874bc2"
                    }
                }
            }
        }, {
            "type": null,
            "id": "f4df1018-b6cb-11e8-b55d-080027874bc2",
            "attributes": {
                "uuid": "f4df1018-b6cb-11e8-b55d-080027874bc2",
                "process_category_uuid": "f4a547b6-b6cb-11e8-bdba-080027874bc2",
                "user_uuid": "f4a4c9f8-b6cb-11e8-8270-080027874bc2",
                "description": "Dolorem quis culpa iste id laborum. Ut illum consequatur facere nam consectetur voluptas. Atque ut omnis ex accusamus quia deserunt suscipit in. Sed consequatur aperiam corporis id eius est.",
                "name": "Iste eos ut qui qui.",
                "status": "INACTIVE",
                "created_at": "2018-09-12 20:39:33",
                "updated_at": "2018-09-12 20:39:33"
            },
            "relationships": {
                "category": {
                    "data": {
                        "type": null,
                        "id": "f4a547b6-b6cb-11e8-bdba-080027874bc2"
                    }
                }
            }
        }, {
            "type": null,
            "id": "f4e0b436-b6cb-11e8-bc0f-080027874bc2",
            "attributes": {
                "uuid": "f4e0b436-b6cb-11e8-bc0f-080027874bc2",
                "process_category_uuid": "f4de7784-b6cb-11e8-ae0b-080027874bc2",
                "user_uuid": "f4de15be-b6cb-11e8-887b-080027874bc2",
                "description": "Voluptatem molestiae beatae sit iusto nobis omnis quasi. Voluptatibus cum in ducimus ipsa pariatur. Dolorem cupiditate magnam impedit alias quibusdam ea et. Perspiciatis at qui aut labore.",
                "name": "Iure aperiam sequi incidunt.",
                "status": "INACTIVE",
                "created_at": "2018-09-12 20:39:33",
                "updated_at": "2018-09-12 20:39:33"
            },
            "relationships": {
                "category": {
                    "data": {
                        "type": null,
                        "id": "f4de7784-b6cb-11e8-ae0b-080027874bc2"
                    }
                }
            }
        }, {
            "type": null,
            "id": "f4dfe39e-b6cb-11e8-a3e6-080027874bc2",
            "attributes": {
                "uuid": "f4dfe39e-b6cb-11e8-a3e6-080027874bc2",
                "process_category_uuid": "f4c1f6cc-b6cb-11e8-8835-080027874bc2",
                "user_uuid": "f4c19088-b6cb-11e8-b7a4-080027874bc2",
                "description": "Sed provident sit maxime minus est. Labore ex odit et commodi quos. Aut qui natus aut adipisci facilis. Totam quia consequatur hic illo aliquam.",
                "name": "Provident rerum sint aut.",
                "status": "INACTIVE",
                "created_at": "2018-09-12 20:39:33",
                "updated_at": "2018-09-12 20:39:33"
            },
            "relationships": {
                "category": {
                    "data": {
                        "type": null,
                        "id": "f4c1f6cc-b6cb-11e8-8835-080027874bc2"
                    }
                }
            }
        }, {
            "type": null,
            "id": "f4dfa0b4-b6cb-11e8-ae03-080027874bc2",
            "attributes": {
                "uuid": "f4dfa0b4-b6cb-11e8-ae03-080027874bc2",
                "process_category_uuid": "f4b8bb02-b6cb-11e8-b450-080027874bc2",
                "user_uuid": "f4b853ba-b6cb-11e8-b41a-080027874bc2",
                "description": "Impedit delectus ad harum reprehenderit incidunt. Aut sed at necessitatibus ducimus soluta. Autem maxime provident dolorem dolor et.",
                "name": "Repellendus dolorem et.",
                "status": "INACTIVE",
                "created_at": "2018-09-12 20:39:33",
                "updated_at": "2018-09-12 20:39:33"
            },
            "relationships": {
                "category": {
                    "data": {
                        "type": null,
                        "id": "f4b8bb02-b6cb-11e8-b450-080027874bc2"
                    }
                }
            }
        }],
    "included": [{
            "type": null,
            "id": "f4d50b36-b6cb-11e8-a5ff-080027874bc2",
            "attributes": {
                "uuid": "f4d50b36-b6cb-11e8-a5ff-080027874bc2",
                "name": "Carmella Mann",
                "status": "ACTIVE",
                "created_at": "2018-09-12 20:39:33",
                "updated_at": "2018-09-12 20:39:33",
                "processes_count": 1
            }
        }, {
            "type": null,
            "id": "f4af3dd4-b6cb-11e8-96f4-080027874bc2",
            "attributes": {
                "uuid": "f4af3dd4-b6cb-11e8-96f4-080027874bc2",
                "name": "Leif Wunsch",
                "status": "ACTIVE",
                "created_at": "2018-09-12 20:39:33",
                "updated_at": "2018-09-12 20:39:33",
                "processes_count": 1
            }
        }, {
            "type": null,
            "id": "f4cb7710-b6cb-11e8-bded-080027874bc2",
            "attributes": {
                "uuid": "f4cb7710-b6cb-11e8-bded-080027874bc2",
                "name": "Mrs. Bert Lockman",
                "status": "ACTIVE",
                "created_at": "2018-09-12 20:39:33",
                "updated_at": "2018-09-12 20:39:33",
                "processes_count": 1
            }
        }, {
            "type": null,
            "id": "f49af20c-b6cb-11e8-9dc7-080027874bc2",
            "attributes": {
                "uuid": "f49af20c-b6cb-11e8-9dc7-080027874bc2",
                "name": "Rebeca Johnson",
                "status": "INACTIVE",
                "created_at": "2018-09-12 20:39:33",
                "updated_at": "2018-09-12 20:39:33",
                "processes_count": 1
            }
        }, {
            "type": null,
            "id": "f4a547b6-b6cb-11e8-bdba-080027874bc2",
            "attributes": {
                "uuid": "f4a547b6-b6cb-11e8-bdba-080027874bc2",
                "name": "Arden Schowalter I",
                "status": "ACTIVE",
                "created_at": "2018-09-12 20:39:33",
                "updated_at": "2018-09-12 20:39:33",
                "processes_count": 1
            }
        }, {
            "type": null,
            "id": "f4de7784-b6cb-11e8-ae0b-080027874bc2",
            "attributes": {
                "uuid": "f4de7784-b6cb-11e8-ae0b-080027874bc2",
                "name": "Mr. Fidel Zemlak II",
                "status": "INACTIVE",
                "created_at": "2018-09-12 20:39:33",
                "updated_at": "2018-09-12 20:39:33",
                "processes_count": 1
            }
        }, {
            "type": null,
            "id": "f4c1f6cc-b6cb-11e8-8835-080027874bc2",
            "attributes": {
                "uuid": "f4c1f6cc-b6cb-11e8-8835-080027874bc2",
                "name": "Jany Gaylord",
                "status": "INACTIVE",
                "created_at": "2018-09-12 20:39:33",
                "updated_at": "2018-09-12 20:39:33",
                "processes_count": 1
            }
        }, {
            "type": null,
            "id": "f4b8bb02-b6cb-11e8-b450-080027874bc2",
            "attributes": {
                "uuid": "f4b8bb02-b6cb-11e8-b450-080027874bc2",
                "name": "Etha Emard",
                "status": "INACTIVE",
                "created_at": "2018-09-12 20:39:33",
                "updated_at": "2018-09-12 20:39:33",
                "processes_count": 1
            }
        }],
    "meta": {
        "total": 8,
        "count": 8,
        "per_page": 15,
        "current_page": 1,
        "total_pages": 1,
        "filter": "INACTIVE",
        "sort_by": "",
        "sort_order": ""
    }
}
```

**TO DO**: Get the type of the model

### CODE REVIEW

File: ProcessMaker/Serializers/ApiSerializer.php [hunk 1]
```diff
@@ -1,71 +1,28 @@
...
-    public function includedData(ResourceInterface $resource, array $data): array
+    protected function getIdFromData(array $data)
     {
-        return $data;
+        if (!array_key_exists('uuid', $data)) {
+            throw new InvalidArgumentException(
+            'JSON API resource objects MUST have a valid id'
+            );
+        }
+        return $data['uuid'];
```
Comment (caleeli) [inline ProcessMaker/Serializers/ApiSerializer.php]: This line loads the **id** field of the response from **uuid**

File: ProcessMaker/Http/Controllers/Api/ProcessController.php [hunk 2]
```diff
@@ -1,20 +1,29 @@
...
      *
+     *
      * @return \Illuminate\Http\Response
      */
-    public function index()
+    public function index(Request $request)
     {
-        //
+        $where = $this->getRequestFilterBy($request, ['name', 'description','status']);
+        $oderBy = $this->getRequestSortBy($request, 'name');
```
Comment (nolanpro) [inline ProcessMaker/Http/Controllers/Api/ProcessController.php]: Typo `oderBy`

File: ProcessMaker/Serializers/ApiSerializer.php (hunk 1 above)
Comment (nolanpro) [reply ProcessMaker/Serializers/ApiSerializer.php]: Where is this method being called from?

**Comment by velkymx** [comment]
Includes is not the keyword in the return, just for the request. The return should be:

also the relationship term should be the name of the child (category in this example) 

category.processes shouldn't be accessible - this is too deep. If you want the process categories with processes, you'd call it via /categories/?id=asldkjasdlkj&include=processes(name of relationship)

File: ProcessMaker/Serializers/ApiSerializer.php (hunk 1 above)
Comment (caleeli) [reply ProcessMaker/Serializers/ApiSerializer.php]: This used internally by Fractal to get the **id** of the model

File: tests/Feature/Api/ProcessControllerTest.php [hunk 3]
```diff
@@ -0,0 +1,196 @@
+<?php
+namespace Tests\Feature\Api;
+
```
[the remark this answers is not in the document]
Comment (caleeli) [reply tests/Feature/Api/ProcessControllerTest.php]: Added DatabaseTransactions and WithFaker. (DatabaseTransactions replaces RefreshDatabase)

File: tests/Feature/Api/ProcessControllerTest.php [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+use Faker\Factory as Faker;
```
[the remark this answers is not in the document]
Comment (caleeli) [reply tests/Feature/Api/ProcessControllerTest.php]: not really, it was removed.

**Pull request merged**